Orders tab: fix order details shown instead of order list in some iPhone devices on iOS 17.0/17.1

jaclync opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #11709
Closes: #11675

<!-- Id number of the GitHub issue this PR addresses. -->

## Why

When split view feature flag is enabled, upon opening the orders tab for the first time, order details are shown instead of the order list in the orders tab in some iPhone devices on iOS 17.0/17.1. It's because `UISplitViewController.isCollapsed` is currently used to check if split view is supported in the UI for an order to be selected, but the value can be `false` in `viewWillAppear` in compact devices on certain iOS versions like iOS 17.0.1 and 17.1.

## Changes

There's a [similar topic in SO](https://stackoverflow.com/questions/44336916/why-sometimes-iscollapsed-of-uisplitviewcontroller-is-false-on-iphone-portrait), and after some testing `UISplitViewController.isCollapsed` becomes `true` in iPhone devices after the first `viewDidLayoutSubviews`. This PR moves the check for selecting an order in split view from `viewWillAppear` to `viewDidLayoutSubviews` once, as `viewDidLayoutSubviews` can be called more than once.

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Hand-testing is recommended. Please try reproducing the issue first.

- In code, enable the `splitViewInOrdersTab` feature flag in `DefaultFeatureFlagService` (returning `true`)
- Launch the app in iOS 17.0.1 or 17.1 simulator/device
- Switch to a store with at least one order, if needed
- Go to the orders tab --> the order list should be shown

### Tablet testing

- Launch the app on a tablet
- Go to the orders tab --> order details should be shown on the right pane while the order list is on the left


## Screenshots
<!-- Include before and after images or gifs when appropriate. -->



https://github.com/woocommerce/woocommerce-ios/assets/1945542/193cff73-9a4b-414d-a890-bafeece9e7ea

https://github.com/woocommerce/woocommerce-ios/assets/1945542/819a4454-8135-47c8-a6e6-50c85a14c4ce



---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync committed:
Move the check for selecting an order in split view from `viewWillAppear` to `viewDidLayoutSubviews` once.

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald commented:
Thanks Jaclyn! I'll test this too, even though I can't repro the issue it's good to have a few eyes on whether it changes anything else.

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -142,6 +142,7 @@ final class OrderListViewController: UIViewController, GhostableViewController {
     ///
     private var noticePresenter: NoticePresenter = DefaultNoticePresenter()
 
+    private var didCheckSelectedItemInSplitView: Bool = false
```
```suggestion
```

Remove if you agree with the suggestion below...

joshheald commented on WooCommerce/Classes/ViewRelated/Orders/OrderListViewController.swift:
```diff
@@ -198,18 +199,26 @@ final class OrderListViewController: UIViewController, GhostableViewController {
+        // The original implementation is to check the selected item in `viewWillAppear` when split view is enabled,
+        // but `UISplitViewController.isCollapsed` can be `false` in `viewWillAppear` in compact devices on certain iOS versions
+        // like iOS 17.0.1 and 17.1.
+        // To fix this issue, the selected item checking is now called once after the first `viewDidLayoutSubviews` where `isCollapsed` value is
+        // correctly set.
+        if !didCheckSelectedItemInSplitView {
+            didCheckSelectedItemInSplitView = true
+            // Select the first order if we're showing in an open split view (i.e. on iPad in some size classes)
+            guard let splitViewController,
+                  !splitViewController.isCollapsed else {
+                return
+            }
+            checkSelectedItem()
+        }
```
```suggestion
        // To fix this issue, the selected item checking is now called after `viewDidLayoutSubviews`, where `isCollapsed` value is
        // correctly set.
        // This additionally ensures that an order is selected when changing from horizontally compact to regular.
        // Select the first order if we're showing in an open split view (i.e. on iPad in some size classes)
        guard let splitViewController,
              !splitViewController.isCollapsed else {
            return
        }
        checkSelectedItem(
```

Not sure about this, so please test it beyond what I've done, but I don't think we need the flag.

Removing it also fixes #11675. However, it might more properly be fixed by adding a call to `checkSelectedItem` to `OrderListViewController.willTransition(to:with:)`.

`willTransition` is called before `viewDidLayoutSubviews`, but both are called on trait change.

joshheald reviewed (commented):
Works well in my testing. However, I think we can make it slightly simpler and fix another issue at the same time; take a look, I'd love your thoughts on the suggested approach.

Leaving as a comment so others will still look at the review to see if they can repro the issue and see it fixed.

jaclync commented:
@bozidarsevo could you also take over this PR / issue when you get a chance? I ran out of time today 🙇🏻‍♀️

Ideally, we can also fix or pave the way for the auto-selection (selecting the first item when none is selected) to work with the deep link issue where the order details presentation clashes with auto-selection in the secondary view (the first known issue in https://github.com/woocommerce/woocommerce-ios/issues/11674 with repro steps). I was thinking we can have two publishers: secondary view state (selected order, deep linked content, or empty state), and collapsed state. Then we can use combineLatest of these two publishers to determine whether to perform auto-selection.

Thanks for the suggestion @joshheald, I was hesitant to always call `checkSelectedItem` in `viewDidLayoutSubviews`/`willTransition` because of the reason above (the deep link issue). Today I was hoping to find another way to get a reliable `isCollapsed` value by trying a few `UISplitViewControllerDelegate` methods and `UISplitViewController.showDetailTargetDidChangeNotification` notification (using a Combine publisher). However, from my testing they all only notify when the view **transitions** but not for the initial accurate state. Maybe `viewDidLayoutSubviews`/`willTransition` is the only way with the current APIs.

bozidarsevo commented:
> the first known issue in #11674 with repro steps

I have a PR opened for that issue 👍 https://github.com/woocommerce/woocommerce-ios/pull/11729

joshheald committed:
11709 Always checkSelectedItem on layout

Removes the “on first time only” flag for checking the selected item.

Checking this for every layout pass resolves issues with size class changes as well.

This may cause issues with deeplinks in future, but we’ll test that.

joshheald committed:
Merge branch 'trunk' into task/11709-check-selected-item-split-view-in-viewDidLayoutSubviews

joshheald commented:
I've added the suggested changes in c7e6e0c, to resolve #11675

As noted in #11731, the fixes here significantly improve UITest behaviour as well, since they suffer from the `isCollapsed` timing issue as well.
